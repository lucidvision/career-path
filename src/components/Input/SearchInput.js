import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { FlatList, StyleSheet, Text, TextInput, TouchableOpacity, View } from 'react-native'
import Fuse from 'fuse.js'
import _ from 'lodash'

class SearchInput extends Component {
  static propTypes = {
    data: PropTypes.array.isRequired,
    onItemPressed: PropTypes.func.isRequired
  }
  state = {
    text: '',
    results: []
  }
  componentDidMount () {
    this.debouncedSearch = _.debounce(this.handleItemSearch, 500)
  }
  handleItemSearch = () => {
    const options = {keys: ['name']}
    const fuse = new Fuse(this.props.data, options)
    const results = fuse.search(this.state.text)
    this.setState({results})
  }
  handleSearchChanged = (text) => {
    this.setState({text})
    this.debouncedSearch()
  }
  handleItemPressed = (item) => {
    this.setState({
      text: '',
      results: []
    })
    this.props.onItemPressed(item)
  }
  renderItem = ({item}) => {
    return (
      <TouchableOpacity onPress={() => this.handleItemPressed(item)} style={styles.listItem}>
        <Text>{item.name}</Text>
      </TouchableOpacity>
    )
  }
  render () {
    return (
      <View style={styles.container}>
        <TextInput
          placeholder='Search'
          onChangeText={this.handleSearchChanged}
          style={styles.input}
          value={this.state.text} />
        {this.state.results.length > 0 &&
          <View style={styles.list}>
            <FlatList
              data={this.state.results}
              keyExtractor={(item) => item.name}
              renderItem={this.renderItem} />
          </View>}
      </View>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    backgroundColor: 'white',
    borderColor: '#d6d7da',
    borderRadius: 4,
    borderWidth: 0.5,
    padding: 10,
    width: '100%',
    zIndex: 1
  },
  input: {
    fontSize: 20
  },
  list: {
    backgroundColor: 'white',
    flex: 1,
    left: 0,
    position: 'absolute',
    right: 0,
    top: 50
  },
  listItem: {
    borderBottomWidth: 0.5,
    borderColor: '#d6d7da',
    height: 40,
    padding: 10
  }
})

export default SearchInput
