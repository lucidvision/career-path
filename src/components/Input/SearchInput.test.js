import React from 'react'
import { shallow } from 'enzyme'
import SearchInput from './SearchInput'

const data = [
  {
    name: 'Doctor',
    salary: 100000,
    educationRequired: 'MD',
    score: 2
  },
  {
    name: 'Nurse',
    salary: 50000,
    educationRequired: 'Bachelors',
    score: 1
  },
  {
    name: 'Therapist',
    salary: 50000,
    educationRequired: 'Bachelors',
    score: 0
  }
]

const createSearchInput = (override) => {
  const props = {
    data: [],
    onItemPressed: () => {},
    ...override
  }

  return shallow(<SearchInput {...props} />)
}

describe('SearchInput', () => {
  const component = createSearchInput()

  it('renders correctly', () => {
    expect(component).toMatchSnapshot()
  })

  describe('handleItemSearch', () => {
    it('sets state with the correct results', () => {
      const component = createSearchInput({data})
      component.setState({text: 't'})
      component.instance().handleItemSearch()
      expect(component.state('results')).toEqual([
        {
          name: 'Therapist',
          salary: 50000,
          educationRequired: 'Bachelors',
          score: 0
        }
      ])
    })
  })

  describe('handleSearchChanged', () => {
    component.instance().debouncedSearch = jest.fn()
    component.instance().handleSearchChanged('Rainbow')

    it('sets state with text', () => {
      expect(component.state('text')).toBe('Rainbow')
    })

    it('calls debouncedSearch function', () => {
      expect(component.instance().debouncedSearch).toHaveBeenCalled()
    })
  })

  describe('handleItemPressed', () => {
    const mockOnItemPressed = jest.fn()
    const component = createSearchInput({onItemPressed: mockOnItemPressed})
    component.setState({text: 'Rainbow', results: [{name: 'Therapist'}]})
    component.instance().handleItemPressed({name: 'Doctor'})

    it('resets the state', () => {
      expect(component.state()).toEqual({text: '', results: []})
    })

    it('calls the onItemPressed callback', () => {
      expect(mockOnItemPressed).toHaveBeenCalled()
    })
  })

  describe('renderItem', () => {
    const renderItem = shallow(component.instance().renderItem({item: {name: 'Doctor'}}))

    it('returns the search result item', () => {
      expect(renderItem.find('Text').props().children).toBe('Doctor')
    })
  })

  describe('when typing into search input', () => {
    it('displays the search results', () => {
      const component = createSearchInput({data})
      component.find('TextInput').props().onChangeText('Doc')
      setTimeout(() => {
        expect(component.state('results')).toEqual([{
          name: 'Doctor',
          salary: 100000,
          educationRequired: 'MD',
          score: 2
        }])
      }, 600)
    })
  })
})
