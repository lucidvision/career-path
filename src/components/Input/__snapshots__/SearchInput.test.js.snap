// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchInput renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "borderColor": "#d6d7da",
      "borderRadius": 4,
      "borderWidth": 0.5,
      "padding": 10,
      "width": "100%",
      "zIndex": 1,
    }
  }
>
  <TextInput
    allowFontScaling={true}
    onChangeText={[Function]}
    placeholder="Search"
    style={
      Object {
        "fontSize": 20,
      }
    }
    value=""
  />
  <View
    style={
      Object {
        "backgroundColor": "white",
        "flex": 1,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 50,
      }
    }
  >
    <FlatList
      data={Array []}
      disableVirtualization={false}
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      numColumns={1}
      onEndReachedThreshold={2}
      renderItem={[Function]}
      scrollEventThrottle={50}
      updateCellsBatchingPeriod={50}
      windowSize={21}
    />
  </View>
</View>
`;
